scalyr_agent/builtin_monitors/linux_process_metrics.py

```diff
@@ -619,163 +814,255 @@ class ProcessMonitor(ScalyrMonitor):
+        # Java agent did this and it is important to some dashboards.
+        for _metric, _metric_value in self.__aggregated_metrics.items():
+            extra = {'app': self.__id}
+            if _metric.type:
+                extra['type'] = _metric.type
+            self._logger.emit_value(_metric.name, _metric_value, extra)
+
+    def __select_processes(self):
+        """Returns a set of the process ids of processes that fulfills the match criteria.
 
         This will either use the commandline matcher or the target pid to find the process.
-        If no process is matched, None is returned.
+        If no process is matched, an empty list is returned.
 
-        @return: The process id of the matching process, or None
-        @rtype: int or None
+        @return: The set of process id(s) of the matching process, or set()
+        @rtype: set()
         """
-        sub_proc = None
-        if self.__commandline_matcher is not None:
-            try:
-                # Spawn a process to run ps and match on the command line.  We only output two
-                # fields from ps.. the pid and command.
-                sub_proc = Popen(['ps', 'ax', '-o', 'pid,command'],
-                                 shell=False, stdout=PIPE)
-
-                sub_proc.stdout.readline()
-                matching_pids = []
-                for line in sub_proc.stdout:
-                    line = line.strip()
-                    if line.find(' ') > 0:
-                        pid = int(line[:line.find(' ')])
-                        line = line[(line.find(' ') + 1):]
-                        if re.search(self.__commandline_matcher, line) is not None:
-                            matching_pids.append( pid )
-
-                total_pids = len( matching_pids )
-                if total_pids == 0:
-                    return None
-                elif total_pids == 1:
-                    return matching_pids[0]
-
-                # if we are here then multiple command lines matched.  First check to
-                # see if self.__pid is in the list and if so return that.  Otherwise
-                # return the first element in the list
-                if self.__pid in matching_pids:
-                    self._logger.warning( "Multiple processes match the command '%s'.  Returning existing pid." % self.__commandline_matcher, limit_once_per_x_secs=300, limit_key='linux-process-monitor-existing-pid' )
-                    return self.__pid
-
-                self._logger.warning( "Multiple processes match the command '%s'.  Returning the pid of the first matching process" % self.__commandline_matcher, limit_once_per_x_secs=300, limit_key='linux-process-monitor-first-pid' )
-
-                return matching_pids[0]
 
-            finally:
-                # Be sure to wait on the spawn process.
-                if sub_proc is not None:
-                    sub_proc.wait()
+        selected_pids = []
```

Comment:
Ok, I think we need to simplify this.  There's too much overlap between `__select_process` and `get_active_matched_pids` and I believe we are missing some functionality here that we require.

Here's how I see the desired logic:
* First, iterate over the current selected pid list.
* For each pid, see if it is still alive.  If it is, stop iterating.  This determine if we have at least one pid still alive that we were last monitoring.
* If we have one pid alive and not `__aggregate_multiple_processes` return the current list of selected pids (which will be just one pid)
* If we have one pid alive and `_aggregate_multiple_processes` then check to `_last_checked_time`.  If it has expired get the set of all matching processes and swap it out for the current set.
* If we do not have one pid alive, then get the set of all matching processes and swap it out for the current set.  (If we are not `__aggregate_multiple_processes`, then emit the log warning if we matched more than one pid and take just the first one.)

Does that make sense?  We can discuss on Monday as needed.